relay/nakama/save.go

```diff
@@ -0,0 +1,190 @@
+package main
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"github.com/heroiclabs/nakama-common/runtime"
+	"github.com/rotisserie/eris"
+)
+
+/*
+	REQUEST MESSAGES
+*/
+
+type SaveGameRequest struct {
+	Data string `json:"data"`
+}
+
+type SaveGameResponse struct {
+	Success bool `json:"success"`
+}
+
+/*
+	STORAGE MODEL
+*/
+
+type Save struct {
+	Data     string `json:"data"`
+	Persona  string `json:"persona"`
+	Verified bool   `json:"verified"`
+}
+
+const (
+	gameSaveCollection = "game_saves"
+)
+
+func initSaveFileStorage(_ runtime.Logger, initializer runtime.Initializer) error {
+	err := initializer.RegisterRpc(
+		"save",
+		handleSaveGame,
+	)
+	if err != nil {
+		return eris.Wrap(err, "")
+	}
+	return nil
+}
+
+func handleSaveGame(ctx context.Context, logger runtime.Logger, _ *sql.DB, nk runtime.NakamaModule, payload string,
+) (string, error) {
+	userID, err := getUserID(ctx)
+	if err != nil {
+		return logErrorMessageFailedPrecondition(logger, eris.Wrap(err, ""), "failed to get user ID")
+	}
+
+	var msg SaveGameRequest
+	err = json.Unmarshal([]byte(payload), &msg)
+	if err != nil {
+		return logError(
+			logger,
+			eris.Wrap(err, `error unmarshalling payload: expected form {"data": <string>}`),
+			InvalidArgument)
+	}
+	// do not allow empty requests
+	if msg.Data == "" {
+		return logErrorFailedPrecondition(
+			logger,
+			eris.New("data cannot be empty"),
+		)
+	}
+
+	// get the persona storage object.
+	persona, err := loadPersonaTagStorageObj(ctx, nk)
+	if err != nil {
+		if eris.Is(eris.Cause(err), ErrPersonaTagStorageObjNotFound) {
+			return "", eris.Wrap(err, "cannot save game: user has not yet claimed a persona tag")
+		}
+		return logErrorFailedPrecondition(logger, eris.Wrap(err, "failed to get persona for save"))
+	}
+	// do not allow saving if they do not yet have an accepted tag.
+	if persona.Status != personaTagStatusAccepted {
+		return logErrorFailedPrecondition(
+			logger,
+			eris.Errorf("persona tag %q is not yet verified: cannot save", persona.PersonaTag),
+		)
+	}
+
+	// check if the user is allowlisted. NOTE: checkVerified will return nil in two cases:
+	// case 1: if the allowlist is disabled (via ENABLE_ALLOWLIST env var).
+	// case 2: the user is actually allowlisted.
+	var verified bool
+	err = checkVerified(ctx, nk, userID)
```

Comment:
What's the reason for including the "verified" field in the game state? 

Above we make sure they have an 'accepted' persona tag status, which means they've made it past the allow-list gate. Is that enough to allow them to save/restore state?

Reply:
i think i confused myself lol. basically the games team needs a way to save progress in the onboarding flow.

such that if they login to nakama with google, enter a beta key, the fact that they entered that should be saved, in the case they don't yet claim a persona, and need to switch to another PC.

i updated the flow to represent the above. lmk how it looks

Reply:
👍